polaris/evaluate/_metric.py

```diff
@@ -116,23 +173,102 @@ def score(
+        # With single-task metrics for multi-task benchmarks, it can happen that some target values are NaN.
+        # We mask these here.
+        mask = mask_index(y_true_targets)
+        kwargs = {"y_true": y_true_targets[mask], self.y_type: pred.mask(mask).flatten()}
         return self.fn(**kwargs, **self.value.kwargs)
 
     def __call__(
-        self, y_true: np.ndarray, y_pred: np.ndarray | None = None, y_prob: np.ndarray | None = None
+        self,
+        y_true: GroundTruth,
+        y_pred: BenchmarkPredictions | None = None,
+        y_prob: BenchmarkPredictions | None = None,
+    ) -> float:
+        return self.score(y_true, y_pred, y_prob)
+
+
+class GroupedMetric(BaseModel):
+    """"""
+
+    metric: Metric
+
+    group_by: str
+    on_error: Literal["ignore", "raise", "default"] = "raise"
+    default: float | None = None
+    aggregation: Literal["mean", "median"] = "mean"
+
+    @field_validator("metric", mode="before")
+    @classmethod
+    def validate_metric(cls, value):
+        if not isinstance(value, Metric):
+            value = Metric[value]
+        return value
+
+    @property
+    def fn(self) -> Callable:
+        """The callable that actually computes the metric"""
+        return self.metric.value.fn
+
+    @property
+    def is_multitask(self) -> bool:
+        """Whether the metric expects a single set of predictions or a dict of predictions."""
+        return self.metric.value.is_multitask
+
+    @property
+    def y_type(self) -> bool:
+        """Whether the metric expects preditive probablities."""
+        return self.metric.value.y_type
+
+    def score(
+        self,
+        y_true: GroundTruth,
+        y_pred: BenchmarkPredictions | None = None,
+        y_prob: BenchmarkPredictions | None = None,
+    ) -> float:
+        pred = check_predictions(y_pred, y_prob, self.y_type)
+
+        df = y_true.as_dataframe()
+        df[self.y_type] = pred.flatten()
+
+        scores = []
+        for _, group in df.groupby(self.group_by):
+            # Because we don't have multi-task metrics yet, we can assume this is a single-task metric
+            y_true_group = group[y_true.target_cols[0]].values
+            y_pred_group = group[self.y_type].values
+
+            # With single-task metrics for multi-task benchmarks, it can happen that some target values are NaN.
+            # We mask these here.
+            mask = mask_index(y_true_group)
+
+            kwargs = {"y_true": y_true_group[mask], self.y_type: y_pred_group[mask]}
+
+            try:
+                score = self.fn(**kwargs, **self.metric.value.kwargs)
+            except Exception as err:
+                if self.on_error == "ignore":
+                    continue
+                elif self.on_error == "raise":
+                    raise
+                elif self.default is not None:
+                    raise ValueError("Default value must be provided when on_error is 'default'") from err
```

Comment:
**Suggestion**
Implement a model validator so that this condition cannot happen:
```python
@model_validator(mode='after')
def check_default_required(self) -> Self:
    handling = self.on_error
    default = self.default
    if handling == 'default' and default is None:
        raise ValueError("Default value must be provided when on_error is 'default'")
    return self
```